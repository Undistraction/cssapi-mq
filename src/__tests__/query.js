import {
  mqWithValidBreakpointsForRange,
  booleanValues,
  genericNumbers,
} from './testHelpers/data'
import cssSerialiser from './helpers/cssSerialiser'

/* eslint-disable ramda/no-redundant-not */

expect.addSnapshotSerializer(cssSerialiser)

describe(`query`, () => {
  it(`throws with no arguments`, () => {
    const mq = mqWithValidBreakpointsForRange(`width`)
    const { query } = mq
    expect(
      () =>
        query()`
        background-color: ${() => `GhostWhite`};
      `
    ).toThrowErrorMatchingSnapshot()
  })

  const invalidValues = [
    ...booleanValues,
    ...genericNumbers,
    null,
    undefined,
    NaN,
    {},
  ]

  for (const value of invalidValues) {
    it(`throws with invalid argument of '${value}'`, () => {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { query } = mq
      expect(
        () =>
          query(value)`
          background-color: ${() => `GhostWhite`};
        `
      ).toThrowErrorMatchingSnapshot()
    })
  }

  for (const value of invalidValues) {
    it(`throws with invalid child of '${value}'`, () => {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { query } = mq
      expect(
        () =>
          query([value])`
          background-color: ${() => `GhostWhite`};
        `
      ).toThrowErrorMatchingSnapshot()
    })
  }

  it(`throws with a nested array`, () => {
    const mq = mqWithValidBreakpointsForRange(`width`)
    const { query } = mq
    expect(
      () =>
        query([[]])`
        background-color: ${() => `GhostWhite`};
      `
    ).toThrowErrorMatchingSnapshot()
  })

  it(`renders query with single feature`, () => {
    const mq = mqWithValidBreakpointsForRange(`width`)
    const { query, minWidth } = mq
    expect(
      query(minWidth(`small`))`
        background-color: ${() => `GhostWhite`};
      `
    ).toMatchSnapshot()
  })

  describe(`and`, () => {
    // @media (min-width: 25em) and (orientation: landscape) {
    it(`renders query with two features anded together`, () => {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { query, minWidth, orientation } = mq
      expect(
        query([minWidth(`small`), orientation(`landscape`)])`
            background-color: ${() => `GhostWhite`};
          `
      ).toMatchSnapshot()
    })

    it(`renders query with multiple features anded together`, () => {
      // @media screen and (min-width: 25em) and (orientation: landscape) and (grid) {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { query, colorGamut, grid, minWidth, orientation } = mq
      expect(
        query([
          colorGamut(`srgb`),
          minWidth(`small`),
          orientation(`landscape`),
          grid(),
        ])`
            background-color: ${() => `GhostWhite`};
          `
      ).toMatchSnapshot()
    })
  })

  describe(`or`, () => {
    it(`renders query with two features ored together`, () => {
      // @media (min-width: 25em),(orientation: landscape) {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { query, minWidth, orientation } = mq
      expect(
        query(minWidth(`small`), orientation(`landscape`))`
            background-color: ${() => `GhostWhite`};
          `
      ).toMatchSnapshot()
    })

    it(`renders query with multiple features ored together`, () => {
      // @media screen,(min-width: 25em),(orientation: landscape),(grid) {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { query, minWidth, orientation, colorGamut, grid } = mq
      expect(
        query(
          colorGamut(`srgb`),
          minWidth(`small`),
          orientation(`landscape`),
          grid()
        )`
            background-color: ${() => `GhostWhite`};
          `
      ).toMatchSnapshot()
    })
  })

  describe(`not`, () => {
    it(`throws with no arguments`, () => {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { query, not } = mq
      expect(
        () =>
          query(not())`
          background-color: ${() => `GhostWhite`};
        `
      ).toThrowErrorMatchingSnapshot()
    })

    for (const value of invalidValues) {
      it(`throws with invalid argument of '${value}'`, () => {
        const mq = mqWithValidBreakpointsForRange(`width`)
        const { query, not } = mq
        expect(
          () =>
            query(not(value))`
            background-color: ${() => `GhostWhite`};
          `
        ).toThrowErrorMatchingSnapshot()
      })
    }

    for (const value of invalidValues) {
      it(`throws with invalid child of '${value}'`, () => {
        const mq = mqWithValidBreakpointsForRange(`width`)
        const { query, not } = mq
        expect(
          () =>
            query(not([value]))`
            background-color: ${() => `GhostWhite`};
          `
        ).toThrowErrorMatchingSnapshot()
      })
    }

    it(`throws with a nested array`, () => {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { query, not } = mq
      expect(
        () =>
          query(not([[]]))`
          background-color: ${() => `GhostWhite`};
        `
      ).toThrowErrorMatchingSnapshot()
    })

    it(`negates anded queries`, () => {
      // @media not screen and (color) and (orientation: landscape) {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { displayMode, query, not, colorGamut, orientation } = mq
      expect(
        query(
          not([
            displayMode(`fullscreen`),
            colorGamut(`p3`),
            orientation(`landscape`),
          ])
        )`
            background-color: ${() => `GhostWhite`};
          `
      ).toMatchSnapshot()
    })

    it(`negates ored queries`, () => {
      // @media not screen, not (color), not (orientation: landscape) {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { displayMode, query, not, colorGamut, orientation } = mq
      expect(
        query(
          not(
            displayMode(`fullscreen`),
            colorGamut(`p3`),
            orientation(`landscape`)
          )
        )`
            background-color: ${() => `GhostWhite`};
          `
      ).toMatchSnapshot()
    })

    describe(`with media type`, () => {
      it(`renders single uquery with media type without adding default media type`, () => {
        // @media (min-width: 25em) {
        const mq = mqWithValidBreakpointsForRange(`width`)
        const { query, not, displayMode, mediaType } = mq
        expect(
          query(not(mediaType(), displayMode(`fullscreen`)))`
              background-color: ${() => `GhostWhite`};
            `
        ).toMatchSnapshot()
      })

      it(`negates anded queries without adding default media type`, () => {
        // @media not screen and (color) and (orientation: landscape) {
        const mq = mqWithValidBreakpointsForRange(`width`)
        const {
          mediaType,
          displayMode,
          query,
          not,
          colorGamut,
          orientation,
        } = mq
        expect(
          query(
            not([
              mediaType(),
              displayMode(`fullscreen`),
              colorGamut(`p3`),
              orientation(`landscape`),
            ])
          )`
              background-color: ${() => `GhostWhite`};
            `
        ).toMatchSnapshot()
      })

      it(`negates ored queries without adding default media type`, () => {
        // @media not screen, not (color), not (orientation: landscape) {
        const mq = mqWithValidBreakpointsForRange(`width`)
        const {
          mediaType,
          displayMode,
          query,
          not,
          colorGamut,
          orientation,
        } = mq
        expect(
          query(
            not(
              mediaType(),
              displayMode(`fullscreen`),
              colorGamut(`p3`),
              orientation(`landscape`)
            )
          )`
              background-color: ${() => `GhostWhite`};
            `
        ).toMatchSnapshot()
      })
    })
  })

  describe(`mixed`, () => {
    it(`allows mixed queries (both and and or)`, () => {
      // @media  screen, (color), screen and (color) and (orientation: landscape) {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { displayMode, query, colorGamut, orientation } = mq
      expect(
        query(displayMode(`fullscreen`), colorGamut(`rec2020`), [
          displayMode(`standalone`),
          orientation(`landscape`),
        ])`
              background-color: ${() => `GhostWhite`};
            `
      ).toMatchSnapshot()
    })

    it(`allows mixed not queries (both and and or)`, () => {
      // @media not screen, not (color), not screen and (color) and (orientation: landscape) {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { displayMode, query, not, colorGamut, orientation, grid } = mq
      expect(
        query(
          not(displayMode(`fullscreen`), colorGamut(`rec2020`), [
            grid(0),
            orientation(`landscape`),
          ])
        )`
              background-color: ${() => `GhostWhite`};
            `
      ).toMatchSnapshot()
    })

    it(`allows mixed queries and not queries (both and and or)`, () => {
      // @media not screen, not (color), not screen and (color) and (orientation: landscape) {
      const mq = mqWithValidBreakpointsForRange(`width`)
      const { grid, atWidth, query, not, colorGamut, orientation } = mq
      expect(
        query(
          grid(),
          colorGamut(`rec2020`),
          [grid(1), orientation(`landscape`)],
          not(grid(), [atWidth(`large`), orientation(`portrait`)])
        )`
              background-color: ${() => `GhostWhite`};
            `
      ).toMatchSnapshot()
    })
  })
})
