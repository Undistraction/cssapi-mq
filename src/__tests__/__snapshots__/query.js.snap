// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query and renders query with multiple features anded together 1`] = `
@media screen and (min-width: 25em) and (orientation: landscape) and (grid) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`query and renders query with two features anded together 1`] = `
@media (min-width: 25em) and (orientation: landscape) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`query mixed allows mixed  queries and not queries (both and and or) 1`] = `
@media screen,(color),screen and (orientation: landscape),not (grid), not (width: 68.75em) and (orientation: portrait) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`query mixed allows mixed not queries (both and and or) 1`] = `
@media not screen, not (color), not screen and (orientation: landscape) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`query mixed allows mixed queries (both and and or) 1`] = `
@media screen,(color),screen and (orientation: landscape) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`query not negates anded queries 1`] = `
@media not screen and (color) and (orientation: landscape) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`query not negates ored queries 1`] = `
@media not screen, not (color), not (orientation: landscape) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`query or renders query with multiple features ored together 1`] = `
@media screen,(min-width: 25em),(orientation: landscape),(grid) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`query or renders query with two features ored together 1`] = `
@media (min-width: 25em),(orientation: landscape) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`query renders query with single feature 1`] = `
@media (min-width: 25em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`query throws with no arguments 1`] = `You must supply at least one argument to 'query()' to build a valid media query`;
